### PR TITLE
Feature/rfc7662 oauth2 token introspection

### PR DESCRIPTION
**Pull request by spack971**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Other, please describe:

---

- [x] You consent that the copyright of your pull request source code belongs to Authlib's author.

Implementation of RFC7662 (#33).

### CODE REVIEW

File: authlib/specs/rfc7662/introspection.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        :class:`authlib.specs.rfc7662.IntrospectionEndpoint`.
+
+        """
+        self.request = request
+        self.query_client = query_client
+
+        self._client = None
+        self._token = None
+
+    def authenticate_revocation_endpoint_client(self):
```
Comment (lepture) [inline authlib/specs/rfc7662/introspection.py]: method name should be `authenticate_ introspection_endpoint_client`

File: authlib/specs/rfc7009/__init__.py [hunk 2]
```diff
@@ -14,3 +14,4 @@
 from .parameters import prepare_revoke_token_request
 from .errors import UnsupportedTokenTypeError
 from .revocation import RevocationEndpoint
+from .registry import *
```
Comment (lepture) [inline authlib/specs/rfc7009/__init__.py]: from .registry import OAUTH_TOKEN_TYPE_HINTS

Comment (lepture) [reply authlib/specs/rfc7009/__init__.py]: I'd rather not make any changes in revocation spec. You can add a `SUPPORTED_TOKEN_TYPES` in IntrospectionEndpoint itself. In this way, people can custom the `SUPPORTED_TOKEN_TYPES` with a subclass.

File: authlib/specs/rfc7662/introspection.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                    return Token.query_by_refresh_token(token, client.client_id)
+                return Token.query_by_access_token(token, client.client_id) or \
+                    Token.query_by_refresh_token(token, client.client_id)
+        """
+        raise NotImplementedError()
+
+    def introspect_token(self, token):
+        """Read given token and return its introspection metadata as a
+        dictionary following RFC7662 keys.
+
```
Comment (lepture) [inline authlib/specs/rfc7662/introspection.py]: remove blank line

File: authlib/specs/rfc7662/introspection.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        )
+        if not self._token:
+            raise InvalidRequestError()
+
+    def create_introspection_response(self):
+        """Validate introspection request and create the response.
+
+        :returns: (status_code, body, headers)
+
+        """
```
Comment (lepture) [inline authlib/specs/rfc7662/introspection.py]: remove blank line

File: authlib/specs/rfc7662/introspection.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    .. _RFC7662: https://tools.ietf.org/html/rfc7662
+
+    """
+    INTROSPECTION_ENDPOINT_AUTH_METHODS = ['client_secret_basic']
+
+    def __init__(self, request, query_client):
+        """Constructor for
+        :class:`authlib.specs.rfc7662.IntrospectionEndpoint`.
+
+        """
```
Comment (lepture) [inline authlib/specs/rfc7662/introspection.py]: no doc string in `__init__`

File: authlib/specs/rfc7662/introspection.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    """Implementation of introspection endpoint which is described in
+    `RFC7662`_.
+
+    :param request: OAuth2Request instance
+    :param query_client: A function to get client by client_id. The client
+        model class MUST implement the methods described by
+        :class:`~authlib.specs.rfc6749.ClientMixin`.
+
+    .. _RFC7662: https://tools.ietf.org/html/rfc7662
+
```
Comment (lepture) [inline authlib/specs/rfc7662/introspection.py]: remove blank line

File: authlib/specs/rfc7662/introspection.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            body = dict(error.get_body())
+            headers = error.get_headers()
+
+        return status, body, headers
+
+    def query_token(self, token, optional=None):
+        """Get the token from database/storage by the given token string.
+        Developers should implement this method::
+
+            def query_token(self, token, token_type_hint, client):
```
Comment (lepture) [inline authlib/specs/rfc7662/introspection.py]: example doesn't match the method interface.

**Comment by lepture** [comment]
I'd rather not make any changes in revocation spec. You can add a SUPPORTED_TOKEN_TYPES in IntrospectionEndpoint itself. In this way, people can custom the SUPPORTED_TOKEN_TYPES with a subclass.

File: authlib/specs/rfc7662/introspection.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        if 'token' not in params:
+            raise InvalidRequestError()
+
+        token_type = params.get('token_type_hint')
+        if token_type and token_type not in self.SUPPORTED_TOKEN_TYPES:
+            raise UnsupportedTokenTypeError()
+
+        self._token = self.query_token(
+            params['token'],
+            {k: v for k, v in params.items() if k != 'token'}
```
Comment (lepture) [inline authlib/specs/rfc7662/introspection.py]: It seems that `query_token` need an client parameter.

**Comment by lepture** [comment]
This will hold on for a while. I'm planning a lot of things in v0.6. It will get merged when I get things ready.

Thank you.

**Comment by spack971** [comment]
OK. Anything else I can do?

**Pull request merged**